File: packages/php/tests/BrowserSessionTest.php
<?php

declare(strict_types=1);

namespace Webpipe\Tests;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Middleware;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;
use Webpipe\Internal\HttpClient;
use Webpipe\WebpipeClient;

final class BrowserSessionTest extends TestCase
{
    private array $history = [];

    private function mockClient(array $responses): WebpipeClient
    {
        $mock = new MockHandler($responses);
        $stack = HandlerStack::create($mock);
        $stack->push(Middleware::history($this->history));
        $guzzle = new GuzzleClient(['handler' => $stack, 'http_errors' => false]);
        $http = HttpClient::create(
            'https://api.web2json.ai',
            'wc-test-key',
            60.0,
            0,
            'webpipe-sdk-php/' . WebpipeClient::VERSION,
            $guzzle,
        );
        return new WebpipeClient(apiKey: 'wc-test-key', httpClient: $http);
    }

    private static function json(array $body, int $status = 200): Response
    {
        return new Response($status, ['Content-Type' => 'application/json'], (string) json_encode($body));
    }

    public function testSessionLifecycle(): void
    {
        $client = $this->mockClient([
            // create
            self::json(['success' => true, 'id' => 'sess-1', 'status' => 'starting']),
            // refresh: starting then running
            self::json(['success' => true, 'id' => 'sess-1', 'status' => 'starting']),
            self::json(['success' => true, 'id' => 'sess-1', 'status' => 'running']),
            // scrape current page
            self::json(['success' => true, 'data' => ['markdown' => '# Hi']]),
            // action
            self::json(['success' => true, 'ok' => true]),
            // delete
            self::json(['success' => true]),
        ]);

        $session = $client->browser->create(url: 'https://example.com', sessionName: 'demo');
        self::assertSame('sess-1', $session->id);
        self::assertSame('starting', $session->status);

        $createBody = json_decode((string) $this->history[0]['request']->getBody(), true);
        self::assertSame('demo', $createBody['session_name']);

        $session->waitUntilRunning(timeout: 5, pollInterval: 0.01);
        self::assertSame('running', $session->status);

        $doc = $session->scrape(formats: ['markdown']);
        self::assertSame('# Hi', $doc->markdown);

        $session->click('button.more', waitAfter: 500);
        $actionBody = json_decode((string) $this->history[4]['request']->getBody(), true);
        self::assertSame(
            ['type' => 'click', 'selector' => 'button.more', 'wait_after' => 500],
            $actionBody,
        );

        $session->close();
        self::assertSame('stopped', $session->status);
        self::assertSame('DELETE', $this->history[5]['request']->getMethod());
    }

    public function testCookies(): void
    {
        $client = $this->mockClient([
            self::json(['success' => true, 'id' => 'sess-2', 'status' => 'running']),
            self::json(['success' => true, 'cookies' => [['name' => 'session', 'value' => 'abc']]]),
            self::json(['success' => true]),
            self::json(['success' => true]),
        ]);

        $session = $client->browser->create();
        $cookies = $session->getCookies();
        self::assertSame('session', $cookies[0]['name']);

        $session->setCookies([['name' => 'x', 'value' => '1', 'domain' => 'example.com']]);
        self::assertSame('set', json_decode((string) $this->history[2]['request']->getBody(), true)['op']);

        $session->clearCookies();
        self::assertSame(['op' => 'clear'], json_decode((string) $this->history[3]['request']->getBody(), true));
    }
}

File: packages/java/src/test/java/ai/webpipe/sdk/BrowserSessionTest.java
package ai.webpipe.sdk;

import static org.junit.jupiter.api.Assertions.*;

import ai.webpipe.sdk.model.Document;
import ai.webpipe.sdk.options.CreateSessionOptions;
import ai.webpipe.sdk.options.NavigateOptions;
import java.util.List;
import org.junit.jupiter.api.Test;

class BrowserSessionTest {

    private WebpipeClient client(MockServer server) {
        return WebpipeClient.builder()
                .apiKey("wc-test-key")
                .baseUrl(server.url())
                .maxRetries(0)
                .build();
    }

    @Test
    void sessionLifecycle() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/browser", 200,
                    "{\"success\":true,\"id\":\"sess-1\",\"status\":\"starting\"}");
            server.enqueue("GET", "/api/v1/browser/sess-1", 200,
                    "{\"success\":true,\"id\":\"sess-1\",\"status\":\"starting\"}");
            server.enqueue("GET", "/api/v1/browser/sess-1", 200,
                    "{\"success\":true,\"id\":\"sess-1\",\"status\":\"running\"}");
            server.enqueue("POST", "/api/v1/browser/sess-1/scrape", 200,
                    "{\"success\":true,\"data\":{\"markdown\":\"# Hi\"}}");
            server.enqueue("POST", "/api/v1/browser/sess-1/action", 200,
                    "{\"success\":true,\"ok\":true}");
            server.enqueue("DELETE", "/api/v1/browser/sess-1", 200, "{\"success\":true}");

            BrowserSession session = client(server).browser.create(
                    new CreateSessionOptions().url("https://example.com").sessionName("demo"));
            assertEquals("sess-1", session.getId());
            assertEquals("starting", session.getStatus());
            assertEquals("demo", server.lastRequest().body().get("session_name").asText());

            session.waitUntilRunning(0.01, 5);
            assertEquals("running", session.getStatus());

            Document doc = session.scrape(new NavigateOptions().formats(List.of("markdown")));
            assertEquals("# Hi", doc.markdown());

            session.click("button.more", 500);
            var actionBody = server.lastRequest().body();
            assertEquals("click", actionBody.get("type").asText());
            assertEquals("button.more", actionBody.get("selector").asText());
            assertEquals(500, actionBody.get("wait_after").asInt());

            session.close();
            assertEquals("stopped", session.getStatus());
            assertEquals("DELETE", server.lastRequest().method());
        }
    }

    @Test
    void cookies() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/browser", 200,
                    "{\"success\":true,\"id\":\"sess-2\",\"status\":\"running\"}");
            server.enqueue("GET", "/api/v1/browser/sess-2/cookies", 200,
                    "{\"success\":true,\"cookies\":[{\"name\":\"session\",\"value\":\"abc\"}]}");
            server.enqueue("POST", "/api/v1/browser/sess-2/cookies", 200, "{\"success\":true}");
            server.enqueue("POST", "/api/v1/browser/sess-2/cookies", 200, "{\"success\":true}");

            BrowserSession session = client(server).browser.create(null);
            assertEquals("session", session.getCookies().get(0).name());

            session.setCookies(List.of(new ai.webpipe.sdk.model.Cookie("x", "1", "example.com", "/")));
            assertEquals("set", server.lastRequest().body().get("op").asText());

            session.clearCookies();
            assertEquals("clear", server.lastRequest().body().get("op").asText());
        }
    }

    @Test
    void asyncSessionLifecycle() throws Exception {
        try (MockServer server = new MockServer()) {
            server.enqueue("POST", "/api/v1/browser", 200,
                    "{\"success\":true,\"id\":\"sess-a1\",\"status\":\"starting\"}");
            server.enqueue("GET", "/api/v1/browser/sess-a1", 200,
                    "{\"success\":true,\"id\":\"sess-a1\",\"status\":\"running\"}");
            server.enqueue("POST", "/api/v1/browser/sess-a1/scrape", 200,
                    "{\"success\":true,\"data\":{\"markdown\":\"# Async\"}}");
            server.enqueue("POST", "/api/v1/browser/sess-a1/action", 200,
                    "{\"success\":true,\"ok\":true}");
            server.enqueue("DELETE", "/api/v1/browser/sess-a1", 200, "{\"success\":true}");

            BrowserSession session = client(server).browser.create(
                    new CreateSessionOptions().url("https://example.com"));

            session.waitUntilRunningAsync(0.01, 5).join();
            assertEquals("running", session.getStatus());

            Document doc = session.scrapeAsync(new NavigateOptions().formats(List.of("markdown"))).join();
            assertEquals("# Async", doc.markdown());

            session.clickAsync("button.more", 500).join();
            assertEquals("click", server.lastRequest().body().get("type").asText());

            session.closeAsync().join();
            assertEquals("stopped", session.getStatus());
        }
    }
}
